### PR TITLE
[SPARK-13899][SQL] Produce InternalRow instead of external Row at CSV data source

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-13899

This PR makes CSV data source produce `InternalRow` instead of `Row`.

Basically, this resembles JSON data source. It uses the same codes for casting.
## How was this patch tested?

Unit tests were used within IDE and code style was checked by `./dev/run_tests`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rxin @falaki

**Comment by HyukjinKwon** [comment]
This PR would allow to infer `TimestampType` more flexibly (e.g. includeing `T` and `GMT`) rather than just using `Timestamp.valueOf()`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -191,12 +192,18 @@ private[csv] object CSVTypeCast {
         case _: DoubleType => Try(datum.toDouble)
           .getOrElse(NumberFormat.getInstance(Locale.getDefault).parse(datum).doubleValue())
         case _: BooleanType => datum.toBoolean
-        case _: DecimalType => new BigDecimal(datum.replaceAll(",", ""))
+        case dt: DecimalType =>
+          val value = new BigDecimal(datum.replaceAll(",", ""))
+          Decimal(value, dt.precision, dt.scale)
         // TODO(hossein): would be good to support other common timestamp formats
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: remove this todo?

**Comment by rxin** [comment]
Thanks - I'm going to merge this. There is a tiny comment. Can you remove that comment in one of your other pr?

**Comment by HyukjinKwon** [comment]
Does that mean closing https://github.com/apache/spark/pull/11550 and not supporting custom date format?

**Pull request closed without merging**